receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/metricsbuilder.go

```diff
@@ -0,0 +1,173 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cwmetricstream // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream"
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/model/pdata"
+	conventions "go.opentelemetry.io/collector/model/semconv/v1.5.0"
+)
+
+const (
+	attributeAWSCloudWatchMetricStreamName = "aws.cloudwatch.metric_stream_name"
+	dimensionInstanceID                    = "InstanceId"
+	namespaceDelimiter                     = "/"
+)
+
+// The resourceMetricsBuilder is used to aggregate metrics for the
+// same metric stream name, account ID, region, and namespace.
+type resourceMetricsBuilder struct {
+	// metricStreamName is the metric stream name.
+	metricStreamName string
+	// accountID is the AWS account ID.
+	accountID string
+	// region is the AWS region.
+	region string
+	// namespace is the CloudWatch metric namespace.
+	namespace string
+	// metricBuilders is the map of metrics within the same
+	// resource group.
+	metricBuilders map[string]*metricBuilder
+}
+
+// newResourceMetricsBuilder creates a resourceMetricsBuilder for the
+// metric stream name, account ID, region, and namespace.
+func newResourceMetricsBuilder(metricStreamName, accountID, region, namespace string) *resourceMetricsBuilder {
+	return &resourceMetricsBuilder{
+		metricStreamName: metricStreamName,
+		accountID:        accountID,
+		region:           region,
+		namespace:        namespace,
+		metricBuilders:   make(map[string]*metricBuilder),
+	}
+}
+
+// AddMetric adds a metric to one of the metric builders based on
+// the key generated for each.
+func (rmb *resourceMetricsBuilder) AddMetric(metric cWMetric) {
+	mb, ok := rmb.metricBuilders[metric.MetricName]
+	if !ok {
+		mb = newMetricBuilder(metric.MetricName, metric.Unit)
+		rmb.metricBuilders[metric.MetricName] = mb
+	}
+	mb.AddDataPoint(metric)
+}
+
+// Build updates the passed in pdata.ResourceMetrics with the metrics in
+// the builder.
+func (rmb *resourceMetricsBuilder) Build(rm pdata.ResourceMetrics) {
+	ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+	rmb.setAttributes(rm.Resource())
+	for _, mb := range rmb.metricBuilders {
+		mb.Build(ilm.Metrics().AppendEmpty())
+	}
+}
+
+// setAttributes creates a pdata.Resource from the fields in the resourceMetricsBuilder.
+// Splits the namespace into service.namespace/service.name if prepended by AWS/.
+func (rmb *resourceMetricsBuilder) setAttributes(resource pdata.Resource) {
+	attributes := resource.Attributes()
+	attributes.InsertString(conventions.AttributeCloudProvider, conventions.AttributeCloudProviderAWS)
+	attributes.InsertString(conventions.AttributeCloudAccountID, rmb.accountID)
+	attributes.InsertString(conventions.AttributeCloudRegion, rmb.region)
+	splitNamespace := strings.SplitN(rmb.namespace, namespaceDelimiter, 2)
```

Comment:
Better to use `Index` and slices instead I think

Reply:
Added `toServiceAttributes` function which uses `Index` and slices.